### PR TITLE
Account and course creation.

### PR DESCRIPTION
**Pull request by pilleye**

### CODE REVIEW

File: app/routes/apiv1/course.js [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    });
+  }
+
+  // Handle the case where the user that is creating a course is an
+  // administrator.
+  if (req.user.is_admin) {
+    return createCourse(req.user, name, description, plan)
+      .then(course => res.status(200).json({
+        success: true,
+        isAdmin: true,
```
Comment (fishdev) [inline app/routes/apiv1/course.js]: See updated API docs regarding coupons, this should be `isFree`.

File: app/routes/apiv1/course.js [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    message: 'Hold tight while we implement this feature.',
+  });
+});
+
+/**
+ * POST /api/course/enroll
+ *
+ * Handles enrolling a student with a UUID.
+ */
+router.post('/enroll', (req, res) => {
```
Comment (fishdev) [inline app/routes/apiv1/course.js]: Consider refactoring routes or at least their handlers into modules.

File: app/routes/auth.js [hunk 3]
```diff
@@ -71,7 +71,7 @@ router.post('/login', (req, res) => {
       sub: user.id,
     };
 
-    const token = jwt.sign(payload, config.auth.secret);
+    const token = jwt.sign(payload, config.auth.secret, { expiresIn: '60d' });
```
Comment (fishdev) [inline app/routes/auth.js]: What is the reasoning behind such a long-lived JWT?

File: app/routes/auth.js [hunk 4]
```diff
@@ -115,6 +117,16 @@ router.post('/signup', (req, res, next) => {
...
 
+  if (!firstName || !lastName) {
+    return res.status(400).json({
+      success: false,
+      error: 'no-name',
+      message: 'Please provide both your first and last name.',
+    });
+  }
+
+  const recaptchaResponse = req.body['g-recaptcha-response'] || '';
```
Comment (fishdev) [inline app/routes/auth.js]: Given the condition on line 130, the conditional assignment here is unnecessary.

File: app/routes/apiv1/course.js [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+  // administrator.
+  if (req.user.is_admin) {
+    return createCourse(req.user, name, description, plan)
+      .then(course => res.status(200).json({
+        success: true,
+        isAdmin: true,
+        courseId: course.uuid,
+      }));
+  }
+
```
Comment (fishdev) [inline app/routes/apiv1/course.js]: Accept a `price` field in the request body and verify that it matches the price on record for the given plan ID to ensure that users cannot see different prices between the creation form and Stripe checkout page.

File: app/routes/apiv1/course.js [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    return res.status(400).json({
+      success: false,
+      error: 'no-plan',
+      message: 'Please select a course plan.',
+    });
+  }
+
+  // Handle the case where the user that is creating a course is an
+  // administrator.
+  if (req.user.is_admin) {
```
Comment (fishdev) [inline app/routes/apiv1/course.js]: Validate the `coupon` received and compute a price relative to whether the user is administrator.

File: app/app.js [hunk 7]
```diff
@@ -1,17 +1,20 @@
 const express = require('express');
 const createError = require('http-errors');
-const path = require('path');
 const cookieParser = require('cookie-parser');
 const logger = require('morgan');
 const passport = require('passport');
-
+const cors = require('cors');
```
Comment (fishdev) [inline app/app.js:10]: In the same vein, is it worth pursuing any form of rate limiting within the server application? Or should this be moved higher up in the stack?

File: app/routes/auth.js [hunk 8]
```diff
@@ -164,6 +179,13 @@ router.post('/signup', (req, res, next) => {
 router.post('/verify', (req, res) => {
   const { email, confirmationId } = req.body;
 
+  if (!email || !confirmationId) {
```
Comment (fishdev) [inline app/routes/auth.js]: How would the client obtain the user's email from a confirmation code? Either:

 - generate unique confirmation IDs and remove the email requirement
 - send the email as a URL parameter to the client confirmation page

**Review by fishdev — CHANGES_REQUESTED** [review]
Looks good overall! General comments:

 - The file structure for routes is sub-optimal. Avoid using utility functions for a single route, instead separate route handlers into their own modules.
 - See updated API docs regarding coupons. The user provides a coupon code in the course creation form, which is validated against their user status. There is an additional endpoint for this at `GET /api/course/create/coupon{?code}`
 - Ensure that dates are indeed being stored relative to UTC. There appears to be some inconsistency in creating instances of `Date` as opposed to using `Date.now()`
 - Fix all linting and code duplication issues from Codacy.

File: app/routes/auth.js (hunk 8 above)
Comment (pilleye) [reply app/routes/auth.js]: I can do that, I just need an email template to send for now.

Comment (fishdev) [reply app/routes/auth.js]: Sending the email as a URL parameter? In that case, we can resolve this for now and I will create an issue on the client for this change.

Comment (pilleye) [reply app/routes/auth.js]: Yeah, I was thinking two URL parameters. Just something like `descartes.quantumstack.xyz/verify?email=aditya@example.com&confirmationId=abc`

File: app/routes/auth.js (hunk 3 above)
Comment (pilleye) [reply app/routes/auth.js]: Why not? I don't know, I guess we need to agree on a common expiration time.

File: app/routes/auth.js (hunk 4 above)
Comment (pilleye) [reply app/routes/auth.js]: Line 138, I `trim()`.

File: app/app.js (hunk 7 above)
Comment (pilleye) [reply app/app.js:10]: Probably just hide behind a Cloudflare Proxy Server.

**Pull request closed without merging**